esp-hal/src/gpio/interrupt.rs

```diff
@@ -0,0 +1,255 @@
+//! force-complete an async operation accidentally from the interrupt handler.
+//!
+//! We could technically use the interrupt status on single-core chips, but it
+//! would be slightly more complicated to prevent the user from breaking things.
+//! (If the user were to clear the interrupt status, we would need to re-enable
+//! it, for PinFuture to detect the completion).
+//!
+//! TODO: currently, direct-binding a GPIO interrupt handler will completely
+//! break the async API. We will need to expose a way to handle async events.
+
+use portable_atomic::{AtomicPtr, Ordering};
+use procmacros::ram;
+use strum::EnumCount;
+
+use crate::{
+    gpio::{GpioBank, InterruptStatusRegisterAccess, asynch, set_int_enable},
+    interrupt::{self, DEFAULT_INTERRUPT_HANDLER, Priority},
+    peripherals::Interrupt,
+    sync::RawMutex,
+};
+
+/// Convenience constant for `Option::None` pin
+pub(super) static USER_INTERRUPT_HANDLER: CFnPtr = CFnPtr::new();
+
+pub(super) static GPIO_LOCK: RawMutex = RawMutex::new();
+
+pub(super) struct CFnPtr(AtomicPtr<()>);
+impl CFnPtr {
+    pub const fn new() -> Self {
+        Self(AtomicPtr::new(core::ptr::null_mut()))
+    }
+
+    pub fn store(&self, f: extern "C" fn()) {
+        self.0.store(f as *mut (), Ordering::Relaxed);
+    }
+
+    pub fn call(&self) {
+        let ptr = self.0.load(Ordering::Relaxed);
+        if !ptr.is_null() {
+            unsafe { (core::mem::transmute::<*mut (), extern "C" fn()>(ptr))() };
+        }
+    }
+}
+
+pub(crate) fn bind_default_interrupt_handler() {
+    // We first check if a handler is set in the vector table.
+    if let Some(handler) = interrupt::bound_handler(Interrupt::GPIO) {
+        let handler = handler as *const unsafe extern "C" fn();
+
+        // We only allow binding the default handler if nothing else is bound.
+        // This prevents silently overwriting RTIC's interrupt handler, if using GPIO.
+        if !core::ptr::eq(handler, DEFAULT_INTERRUPT_HANDLER.handler() as _) {
+            // The user has configured an interrupt handler they wish to use.
+            info!("Not using default GPIO interrupt handler: already bound in vector table");
+            return;
+        }
+    }
+
+    // The vector table doesn't contain a custom entry. Still, the
+    // peripheral interrupt may already be bound to something else.
+    for cpu in cores() {
+        if interrupt::bound_cpu_interrupt_for(cpu, Interrupt::GPIO).is_some() {
+            info!("Not using default GPIO interrupt handler: peripheral interrupt already in use");
+            return;
+        }
+    }
+
+    unsafe { interrupt::bind_interrupt(Interrupt::GPIO, default_gpio_interrupt_handler) };
+
+    // By default, we use lowest priority
+    set_interrupt_priority(Interrupt::GPIO, Priority::min());
+}
+
+cfg_if::cfg_if! {
+    if #[cfg(esp32)] {
+        // On ESP32, the interrupt fires on the core that started listening for a pin event.
```

Comment:
This also means we technically don't need a RawMutex, but a simple interrupt-free context on ESP32, but even this way ESP32 can be less intrusive on the other core than if we were to just route interrupts to core 0. With that said, I think this is enough special-handling for the ESP32. We can maybe return to this question with the P4.